platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileReaderPbj.java

```diff
@@ -0,0 +1,465 @@
+     *
+     * @return True if file is open for reading
+     */
+    public boolean isOpen() {
+        return open.get();
+    }
+
+    /** Close this data file, it can not be used once closed. */
+    public void close() throws IOException {
+        open.set(false);
+        for (int i = 0; i < MAX_FILE_CHANNELS; i++) {
+            final FileChannel fileChannel = fileChannels.getAndSet(i, null);
+            if (fileChannel != null) {
+                fileChannel.close();
+            }
+        }
+    }
+
+    // =================================================================================================================
+    // Private methods
+
+    /**
+     * Opens a new file channel for reading the file, if the total number of channels opened is
+     * less than {@link #MAX_FILE_CHANNELS}. This method is safe to call from multiple threads.
+     *
+     * @param index Index of the new file channel. If greater or equal to {@link
+     *                            #MAX_FILE_CHANNELS}, no new channel is opened
+     * @throws IOException
+     *      If an I/O error occurs
+     */
+    protected void openNewFileChannel(final int index) throws IOException {
+        if (index >= MAX_FILE_CHANNELS) {
+            return;
+        }
+        final FileChannel fileChannel = FileChannel.open(path, StandardOpenOption.READ);
+        if (fileChannels.compareAndSet(index, null, fileChannel)) {
+            fileChannelsCount.incrementAndGet();
+        } else {
+            fileChannel.close();
+        }
+    }
+
+    /**
+     * Replaces a closed file channel at a given index in {@link #fileChannels} with a new one.
+     * This method is safe to be called from multiple threads. If a channel is closed, and two
+     * threads are calling this method to replace it with a new channel, only one of them will
+     * proceed, while the channel opened in the other thread will be closed immediately and
+     * not used any further.
+     *
+     * @param index File channel index in {@link #fileChannels}
+     * @param closedChannel Closed file channel to replace
+     * @throws IOException
+     *      If an I/O error occurs
+     */
+    protected void reopenFileChannel(final int index, final FileChannel closedChannel) throws IOException {
+        assert index < MAX_FILE_CHANNELS;
+        // May be closedChannel or may be already reopened in a different thread
+        assert fileChannels.get(index) != null;
+        assert !closedChannel.isOpen();
+        final FileChannel fileChannel = FileChannel.open(path, StandardOpenOption.READ);
+        if (!fileChannels.compareAndSet(index, closedChannel, fileChannel)) {
+            fileChannel.close();
+        }
+    }
+
+    /**
+     * Returns an index of an opened file channel to read data. Opens a new file channel, if
+     * possible, when the number of file channels currently in use is much greater than the number
+     * of opened file channels.
```

Comment:
This is a bit confusing due to poor word choices.  Perhaps rewrite as follows:
```suggestion
     * Returns an index of an opened file channel to read data and increments the lease count.
     * Opens a new file channel, if possible, when the lease count per channel is greater than {@value #THREADS_PER_FILECHANNEL}.
```

Reply:
✅